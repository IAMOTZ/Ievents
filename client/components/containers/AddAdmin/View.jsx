import React from 'react';
import PropTypes from 'prop-types';
import SideNavigation from '../../common/SideNavigation';
import Header from '../../common/Header';
import Footer from '../../common/Footer';
import { AuthTopNavigation } from '../../common/TopNavigation';
import { LoadingIcon } from '../../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../../common/Alert';

const View = props => (
  <div id="add-admin-container">
    {/* Top navigation on small screen */}
    <SideNavigation
      name={props.userName}
      isAdmin={props.isAdmin}
      isSuperAdmin={props.isSuperAdmin}
      dispatch={props.dispatch}
    />
    <div className="container-fluid">
      <div className="row">
        {/*  Side navigation on large screen */}
        <AuthTopNavigation
          name={props.userName}
          title="Add an admin"
          isAdmin={props.isAdmin}
          isSuperAdmin={props.isSuperAdmin}
          dispatch={props.dispatch}
        />
        <div className="col-lg-10 offset-md-2 mt-lg-0">
          <Header text="Add an admin" />
          <div className="page-content">
            <div className="d-flex flex-column align-items-center mt-5">
              <LoadingIcon start={props.addingAdminStarted} size={3} />
              <div className="card align-items-center text-center w-75">
                <div className="card-body">
                  <BigAlert message={props.addingAdminError} />
                  {
                    props.addingAdminResolved ?
                      <BigAlert
                        message={`${props.newAdmin} is now an admin`}
                        type="success"
                      /> : null
                  }
                  <div className="input-group px-3">
                    <span className="input-group-addon">
                      <i className="fa fa-envelope" />
                    </span>
                    <input
                      id="input-email"
                      type="text"
                      className="form-control"
                      placeholder="The user's email"
                      name="email"
                      onChange={props.getInput}
                    />
                  </div>
                  <SmallAlert message={props.inputErrors.emailError} />
                  <ul className="text-left mt-2">
                    <li>The user must have signed up already</li>
                    <li>If successful, the user would have access to:
                      <ol type="number">
                        <li>Creating a Center</li>
                        <li>Editing a Center</li>
                        <li>Viewing transactions</li>
                        <li>Making decisions on transactions</li>
                      </ol>
                    </li>
                  </ul>
                  <button
                    id="add-btn"
                    className="btn ie-blue-button"
                    disabled={props.addingAdminStarted}
                    onClick={props.add}
                  >Add Admin
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <span className="d-block d-sm-none mt-5">
      <Footer />
    </span>
  </div>
);


View.defaultProps = {
  newAdmin: '',
  addingAdminError: '',
  inputErrors: {},
};

/* eslint-disable react/forbid-prop-types */
View.propTypes = {
  userName: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  isSuperAdmin: PropTypes.bool.isRequired,
  inputErrors: PropTypes.object,
  addingAdminStarted: PropTypes.bool.isRequired,
  addingAdminError: PropTypes.string,
  addingAdminResolved: PropTypes.bool.isRequired,
  newAdmin: PropTypes.string,
  getInput: PropTypes.func.isRequired,
  add: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default View;
