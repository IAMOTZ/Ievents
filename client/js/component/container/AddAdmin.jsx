import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import { addAdmin, clearStatus } from '../../actions/authAction';

import UserSideNav from '../common/SideNavigation.jsx';
import Header from '../common/Header.jsx';
import { UserTopNav } from '../common/TopNavigation.jsx';

@connect((store) => {
  return {
    user: store.user.user,
    authenticated: store.user.status.fetched,
    status: {
      success: store.user.status.adminAdded,
      error: store.user.status.addingAdminError.message
    }
  }
})

export default class AddAdmin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      inputError: null,
    }
  }

  refresh() {
    this.props.dispatch(clearStatus('ADD_ADMIN'));
  }

  getInput = (e) => {
    if (this.props.status.success) {
      this.refresh();
    }
    const state = this.state;
    state[e.target.name] = e.target.value;
    this.setState(state);
  }

  add = (e) => {
    if (this.state.email) {
      this.setState({ inputError: null });
      this.props.dispatch(addAdmin(this.state.email, this.props.user.token));
    } else {
      this.setState({ inputError: 'Email is required' });
    }
  }

  render() {
    if (!this.props.authenticated) {
      return (<Redirect to="/users/login" />);
    } else {
      return (
        <div id="add-admin-container">
          {/* Top navigation on small screen */}
          <UserTopNav name={this.props.user.name} title='Add an admin' />
          <div className="container-fluid">
            <div className="row">
              {/*  Side navigation on large screen */}
              <UserSideNav userName={this.props.user.name} />
              {/* Main content */}
              <div className="col-lg-10 offset-md-2 mt-lg-0" id="main-content">
                {/* Content Header(navigation) on large screen */}
                <Header text='Add an admin' />
                <div className="card text-center align-items-center mt-5 w-75 mx-auto">
                  <div className="card-body">
                    <Alert
                      inputError={this.state.inputError}
                      addingError={this.props.status.error}
                      success={this.props.status.success}
                      newAmdin={this.state.email} />
                    <div className="input-group px-3">
                      <span className="input-group-addon">
                        <i className="fa fa-user"></i>
                      </span>
                      <input
                        type="text"
                        className="form-control"
                        placeholder="The user's email"
                        name="email"
                        onChange={this.getInput} />
                    </div>
                    <ul className="text-left mt-2 text-muted">
                      <li>The user must have signed up already</li>
                      <li>If successful, the user would have access to:
                        <ol type="number">
                          <li>Creating a Center</li>
                          <li>Editing a Center</li>
                          <li>Viewing transactions</li>
                          <li>Making decisions on transactions</li>
                        </ol>
                      </li>
                    </ul>
                    <button
                      className="btn btn-primary"
                      onClick={this.add}>Add Admin</button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <footer className="d-block d-sm-none mt-5">
            <div className="container text-white text-center py-5">
              <h1>Ievents</h1>
              <p>Copyright &copy; 2017</p>
            </div>
          </footer>
        </div>
      );
    }
  }
}

const Alert = (props) => {
  if (props.addingError || props.inputError) {
    return (
      <p className="text-danger">{props.inputError || props.addingError}</p>
    );
  } else if (props.success) {
    return (
      <span>
        <p className="text-success">Admin Added!</p>
        <span>{props.newAmdin} is now and admin</span>
      </span>
    );
  } else {
    return null;
  }
}
