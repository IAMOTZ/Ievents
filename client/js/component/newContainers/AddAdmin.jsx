import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { validateAddAdminInputs } from '../../helpers/inputValidators';
import { addAdmin } from '../../actions/authAction';
import { stopAsyncProcess } from '../../actions/commonActions';
import * as asyncProcess from '../../actions/asyncProcess';
import UserSideNav from '../common/SideNavigation';
import Header from '../common/Header';
import { UserTopNav } from '../common/TopNavigation';
import { LoadingIcon } from '../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../common/Alert';

@connect((store) => {
  const { user } = store.authReducer;
  return {
    user,
    isAdmin: user.role === 'admin',
    isSuperAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isUserAuthenticaed: store.authReducer.loggingUserResolved,
    addingAdminStarted: store.addAdminReducer.addingAdminStarted,
    addingAdminResolved: store.addAdminReducer.addingAdminResolved,
    addingAdminError: store.addAdminReducer.addingAdminError,
  };
})
class AddAdmin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      inputErrors: {
        emailError: null,
      },
    };
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_ADMIN));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    if (this.props.addingAdminResolved || this.props.addingAdminError) {
      this.refresh();
    }
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Clears all the inputErrors in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      emailError: null,
    };
    this.setState(state);
  }

  /**
   * Refresh the page by stopping any asynchronous process that might be going on.
   */
  refresh() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_ADMIN));
  }

  /**
   * Dispatches the action to add the admin.
   */
  add = () => {
    const { email } = this.state;
    const inputErrors = validateAddAdminInputs({ email });
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(addAdmin(this.state.email, this.props.user.token));
    }
  }

  render() {
    if (!this.props.isUserAuthenticaed) {
      return (<Redirect to="/users/login" />);
    }
    return (
      <div id="add-admin-container">
        {/* Top navigation on small screen */}
        <UserTopNav
          name={this.props.user.name}
          title="Add an admin"
          isAdmin={this.props.isAdmin}
          isSuperAdmin={this.props.isSuperAdmin}
          dispatch={this.props.dispatch}
        />
        <div className="container-fluid">
          <div className="row">
            {/*  Side navigation on large screen */}
            <UserSideNav
              name={this.props.user.name}
              isAdmin={this.props.isAdmin}
              isSuperAdmin={this.props.isSuperAdmin}
              dispatch={this.props.dispatch}
            />
            {/* Main content */}
            <div className="col-lg-10 offset-md-2 mt-lg-0" id="main-content">
              {/* Content Header(navigation) on large screen */}
              <Header text="Add an admin" />
              <div className="d-flex flex-column align-items-center mt-5">
                <LoadingIcon start={this.props.addingAdminStarted} size={3} />
                <div className="card align-items-center text-center w-75">
                  <div className="card-body">
                    <BigAlert message={this.props.addingAdminError} />
                    {
                      this.props.addingAdminResolved ?
                        <BigAlert
                          message={`${this.state.email} is now an admin`}
                          type="success"
                        /> : null
                    }
                    <div className="input-group px-3">
                      <span className="input-group-addon">
                        <i className="fa fa-user" />
                      </span>
                      <input
                        id="input-email"
                        type="text"
                        className="form-control"
                        placeholder="The user's email"
                        name="email"
                        onChange={this.getInput}
                      />
                    </div>
                    <SmallAlert message={this.state.inputErrors.emailError} />
                    <ul className="text-left mt-2 text-muted">
                      <li>The user must have signed up already</li>
                      <li>If successful, the user would have access to:
                          <ol type="number">
                          <li>Creating a Center</li>
                          <li>Editing a Center</li>
                          <li>Viewing transactions</li>
                          <li>Making decisions on transactions</li>
                        </ol>
                      </li>
                    </ul>
                    <button
                      id="add-btn"
                      className="btn btn-primary pointer-button"
                      disabled={this.props.addingAdminStarted}
                      onClick={this.add}
                    >Add Admin
                      </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer className="d-block d-sm-none mt-5">
          <div className="container text-white text-center py-5">
            <h1>Ievents</h1>
            <p>Copyright &copy; 2017</p>
          </div>
        </footer>
      </div>
    );
  }
}

export default AddAdmin;
