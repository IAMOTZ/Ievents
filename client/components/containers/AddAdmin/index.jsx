import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { validateAddAdminInputs } from '../../../helpers/inputValidators';
import { addAdmin } from '../../../actions/userActions';
import { stopAsyncProcess } from '../../../actions/commonActions';
import * as asyncProcess from '../../../actions/asyncProcess';
import View from './View';

@connect((store) => {
  const { user } = store.authReducer;
  return {
    userName: user.name,
    userToken: user.token,
    isAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isSuperAdmin: user.role === 'superAdmin',
    addingAdminStarted: store.addAdminReducer.addingAdminStarted,
    addingAdminResolved: store.addAdminReducer.addingAdminResolved,
    addingAdminError: store.addAdminReducer.addingAdminError,
  };
})
class AddAdmin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      inputErrors: {
        emailError: null,
      },
    };
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_ADMIN));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    if (this.props.addingAdminResolved || this.props.addingAdminError) {
      this.refresh();
    }
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Clears all the inputErrors in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      emailError: null,
    };
    this.setState(state);
  }

  /**
   * Refresh the page by stopping any asynchronous process that might be going on.
   */
  refresh() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.ADDING_ADMIN));
  }

  /**
   * Dispatches the action to add the admin.
   */
  add = () => {
    const { email } = this.state;
    const inputErrors = validateAddAdminInputs({ email });
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(addAdmin(this.state.email, this.props.userToken));
    }
  }

  render() {
    return (
      <View
        userName={this.props.userName}
        isAdmin={this.props.isAdmin}
        isSuperAdmin={this.props.isSuperAdmin}
        dispatch={this.props.dispatch}
        addingAdminStarted={this.props.addingAdminStarted}
        addingAdminError={this.props.addingAdminError}
        addingAdminResolved={this.props.addingAdminResolved}
        newAdmin={this.state.email}
        getInput={this.getInput}
        inputErrors={this.state.inputErrors}
        add={this.add}
      />
    );
  }
}

AddAdmin.defaultProps = {
  userName: '',
  userToken: '',
  isAdmin: false,
  isSuperAdmin: false,
  addingAdminStarted: false,
  addingAdminResolved: false,
  addingAdminError: '',
  dispatch: () => {},
};

AddAdmin.propTypes = {
  userName: PropTypes.string,
  userToken: PropTypes.string,
  isAdmin: PropTypes.bool,
  isSuperAdmin: PropTypes.bool,
  addingAdminStarted: PropTypes.bool,
  addingAdminResolved: PropTypes.bool,
  addingAdminError: PropTypes.string,
  dispatch: PropTypes.func,
};

export default AddAdmin;
