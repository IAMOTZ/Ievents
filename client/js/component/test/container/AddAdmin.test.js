/* global shallow */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import AddAdmin from '../../container/AddAdmin.jsx';

const props = {
  user: { name: 'test-user' },
  authenticated: true,
  status: {
    adding: false,
    success: false,
    error: false,
  },
  dispatch: () => { },
};
const alterProps = newProps => ({
  user: newProps.user || props.user,
  authenticated: newProps.authenticated === undefined ? true : newProps.authenticated,
  status: {
    adding: newProps.adding || false,
    success: newProps.success || false,
    error: newProps.error || false,
  },
  dispatch: () => { },
});

describe('<AddAdmin />', () => {
  describe('Rendering:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<AddAdmin {...props} />);
    });
    it('should redirect to login page if user is not authenticated', () => {
      wrapper.setProps(alterProps({ authenticated: false }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly on page load', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when adding', () => {
      wrapper.setProps(alterProps({ adding: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Behaviour:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<AddAdmin {...props} />);
    });
    it('should update the email state as the user inputs email', () => {
      let email = 'test@';
      wrapper
        .find('#input-email')
        .simulate('change', { target: { name: 'email', value: email } });
      expect(wrapper.state('email')).toEqual(email);
      email = 'test@gmail.com';
      wrapper
        .find('#input-email')
        .simulate('change', { target: { name: 'email', value: email } });
      expect(wrapper.state('email')).toEqual(email);
    });
    it('should update inputError state when user tries to add without email', () => {
      wrapper.find('#add-btn').simulate('click');
      expect(wrapper.state('inputError')).toEqual('Email is required');
    });
    it('should clear inputError state when user tries to add with email', () => {
      wrapper.setState({ inputError: 'Just any error' });
      wrapper
        .find('#input-email')
        .simulate('change', { target: { name: 'email', value: 'test@gmail.com' } });
      wrapper.find('#add-btn').simulate('click');
      expect(wrapper.state('inputError')).toBeNull();
    });
  });
});
