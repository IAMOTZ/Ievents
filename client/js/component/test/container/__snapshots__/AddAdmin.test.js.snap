// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddAdmin /> Rendering: should redirect to login page if user is not authenticated 1`] = `
<Redirect
  push={false}
  to="/users/login"
/>
`;

exports[`<AddAdmin /> Rendering: should render correctly on page load 1`] = `
<div
  id="add-admin-container"
>
  <UserTopNav
    dispatch={[Function]}
    name="test-user"
    title="Add an admin"
  />
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <UserSideNav
        dispatch={[Function]}
        name="test-user"
      />
      <div
        className="col-lg-10 offset-md-2 mt-lg-0"
        id="main-content"
      >
        <Header
          text="Add an admin"
        />
        <div
          className="d-flex flex-column align-items-center mt-5"
        >
          <LoadingIcon
            size={3}
            start={false}
          />
          <div
            className="card align-items-center text-center w-75"
          >
            <div
              className="card-body"
            >
              <BigAlert />
              <div
                className="input-group px-3"
              >
                <span
                  className="input-group-addon"
                >
                  <i
                    className="fa fa-user"
                  />
                </span>
                <input
                  className="form-control"
                  id="input-email"
                  name="email"
                  onChange={[Function]}
                  placeholder="The user's email"
                  type="text"
                />
              </div>
              <SmallAlert
                message={null}
              />
              <ul
                className="text-left mt-2 text-muted"
              >
                <li>
                  The user must have signed up already
                </li>
                <li>
                  If successful, the user would have access to:
                  <ol
                    type="number"
                  >
                    <li>
                      Creating a Center
                    </li>
                    <li>
                      Editing a Center
                    </li>
                    <li>
                      Viewing transactions
                    </li>
                    <li>
                      Making decisions on transactions
                    </li>
                  </ol>
                </li>
              </ul>
              <button
                className="btn btn-primary pointer-button"
                disabled={false}
                id="add-btn"
                onClick={[Function]}
              >
                Add Admin
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="d-block d-sm-none mt-5"
  >
    <div
      className="container text-white text-center py-5"
    >
      <h1>
        Ievents
      </h1>
      <p>
        Copyright © 2017
      </p>
    </div>
  </footer>
</div>
`;

exports[`<AddAdmin /> Rendering: should render correctly when adding 1`] = `
<div
  id="add-admin-container"
>
  <UserTopNav
    dispatch={[Function]}
    name="test-user"
    title="Add an admin"
  />
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <UserSideNav
        dispatch={[Function]}
        name="test-user"
      />
      <div
        className="col-lg-10 offset-md-2 mt-lg-0"
        id="main-content"
      >
        <Header
          text="Add an admin"
        />
        <div
          className="d-flex flex-column align-items-center mt-5"
        >
          <LoadingIcon
            size={3}
            start={true}
          />
          <div
            className="card align-items-center text-center w-75"
          >
            <div
              className="card-body"
            >
              <BigAlert />
              <div
                className="input-group px-3"
              >
                <span
                  className="input-group-addon"
                >
                  <i
                    className="fa fa-user"
                  />
                </span>
                <input
                  className="form-control"
                  id="input-email"
                  name="email"
                  onChange={[Function]}
                  placeholder="The user's email"
                  type="text"
                />
              </div>
              <SmallAlert
                message={null}
              />
              <ul
                className="text-left mt-2 text-muted"
              >
                <li>
                  The user must have signed up already
                </li>
                <li>
                  If successful, the user would have access to:
                  <ol
                    type="number"
                  >
                    <li>
                      Creating a Center
                    </li>
                    <li>
                      Editing a Center
                    </li>
                    <li>
                      Viewing transactions
                    </li>
                    <li>
                      Making decisions on transactions
                    </li>
                  </ol>
                </li>
              </ul>
              <button
                className="btn btn-primary pointer-button"
                disabled={true}
                id="add-btn"
                onClick={[Function]}
              >
                Add Admin
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="d-block d-sm-none mt-5"
  >
    <div
      className="container text-white text-center py-5"
    >
      <h1>
        Ievents
      </h1>
      <p>
        Copyright © 2017
      </p>
    </div>
  </footer>
</div>
`;
